### DISCUSSION TITLE
Expected Behavior When Event-level Data has 0 Rows

### DISCUSSION THREAD

**Discussion by mattroumaya** [Gilead-Public/gsm#514] [General]
I noticed the following when writing a new unit test for `RunAssessment()`. The Serious Adverse Event Assessment contains two `FilterDomain()` steps, and for the minimal testing data found in `testthat`, it wipes out `dfAE` so that it contains 0 rows. 
- There is a warning that it was filtered to 0 rows, so that is fine. 

The real question I have is if the following should still run and produce a result, or if we should build in a stop that doesn't return any KRIs since all AE data was filtered out:

### 1. This example runs without errors when `dfAE` is a `tibble`
```r
dfAE <- tibble::tibble(
  SubjectID = character(0),
  AE_SERIOUS = character(0),
  AE_TE_FLAG = logical(0),
  AE_GRADE = numeric(0)
  )

dfSUBJ <- data.frame(
  stringsAsFactors = FALSE,
         SubjectID = c("1234", "5678", "9876"),
            SiteID = c("X010X", "X102X", "X999X"),
   TimeOnTreatment = c(3455, 1745, 1233),
       TimeOnStudy = c(1234, 2345, 4567),
          RandDate = c("2012-09-02", "2017-05-08", "2018-05-20")
)

AE_Map_Raw(dfs = list(dfAE = dfAE, dfSUBJ = dfSUBJ))
```

### 2. This example errors out when `dfAE` is a `data.frame`

Interestingly, the below returns `NULL` (due to failing `is_mapping_valid()`) when `dfAE` is a `data.frame` and not a `tibble`:

```r
dfAE <- data.frame(
  stringsAsFactors = FALSE,
         SubjectID = c(),
        AE_SERIOUS = c(),
        AE_TE_FLAG = c(),
          AE_GRADE = c()
)

dfSUBJ <- data.frame(
  stringsAsFactors = FALSE,
         SubjectID = c("1234", "5678", "9876"),
            SiteID = c("X010X", "X102X", "X999X"),
   TimeOnTreatment = c(3455, 1745, 1233),
       TimeOnStudy = c(1234, 2345, 4567),
          RandDate = c("2012-09-02", "2017-05-08", "2018-05-20")
)

AE_Map_Raw(dfs = list(dfAE = dfAE, dfSUBJ = dfSUBJ))
```

### 3. Initial example where I noticed this
- See `sae_assessment$lData$dfAE` and `sae_assessment$lResults`

```r
dfAE <- data.frame(
  stringsAsFactors = FALSE,
         SubjectID = c("1234", "1234", "5678", "5678"),
        AE_SERIOUS = c("No", "No", "Yes", "No"),
        AE_TE_FLAG = c(TRUE, TRUE, FALSE, FALSE),
          AE_GRADE = c(1, 3, 1, 4)
)

dfSUBJ <- data.frame(
  stringsAsFactors = FALSE,
         SubjectID = c("1234", "5678", "9876"),
            SiteID = c("X010X", "X102X", "X999X"),
   TimeOnTreatment = c(3455, 1745, 1233),
       TimeOnStudy = c(1234, 2345, 4567),
          RandDate = c("2012-09-02", "2017-05-08", "2018-05-20")
)

lAssessments <- MakeAssessmentList()
  lData <- list(
    dfAE = dfAE,
    dfSUBJ = dfSUBJ
  )
  lTags <- list(
    Study = "myStudy"
  )
  lMapping <- clindata::mapping_rawplus

  sae_assessment <- RunAssessment(lAssessments$sae, lData = lData, lMapping = lMapping, lTags = lTags)
```

**Comment by mattroumaya** [comment]
Tagging @jwildfire @gwu05 @samussiah for you to take a look at this and weigh-in when you have the chance.

**Comment by jwildfire** [comment]
Nice write-up @mattroumaya - appreciate the reprexes.

I think that the behavior in examples 1 & 3 above is preferable. It's definitely possible that there will be times early in a study when 0 AEs have been reported, so I don't think that event should trigger an error that stops the workflow. 

The bigger issue here is really that the stat assessment is very likely invalid when all of the rates are 0. We need to do some more work to make sure that statistical assumptions are met, and to throw informative warnings (or even errors) when they aren't met. This is somewhat related to #463 in that it might make sense to fall back to a non-statistical workflow when stat assumptions aren't met. 

For now, I'm going to go ahead and file an issue to make sure we dig in to the stat assumptions a bit more and make sure that informative messages are added in the `X_Analyze` functions. 

@gwu05 - thinking you and @ge-li might be able to take the lead on this as part of v1.1 or v1.2.

**Comment by jwildfire** [comment]
@mattroumaya I'm leaning towards just leaving the core behavior the same here since I'm not sure if there's an easy way to avoid the behavior in example 2 above. It's not a huge deal, but also not ideal to have somewhat arbitrary differences in behavior between tibbles and data.frames if that can be avoided. I guess we could just require tibbles throughout, but that seems like overkill ...
